src/Arrays/buffers.jl

```diff
@@ -0,0 +1,165 @@
+module Buffers
+
+using Requires
+using GPUifyLoops
+using MPI
+
+import ..haspkg
+
+##
+# Implements a double buffering for MPI communication
+# MPI.Isend and MPI.Irecv operate on the secondary buffers
+# and the primary buffers are used as a staging ground to
+# scatter the data from. 
+#
+# Status:
+# - Unoptimized host memory
+# - Pinned host memory
+#
+# TODO:
+# - Optimize this scheme for MPI-aware CUDA
+#   should remove one of the buffers
+# - Implement HostBuffer
+# - Implement DeviceBuffer
+
+export DoubleBuffer, BufferKind
+
+@enum BufferKind begin
+  Host
+  Pinned
+  HostBuffer
+  DeviceBuffer
+end
+
+struct DoubleBuffer{T, Arr, Buff}
+  primary   :: Arr
+  secondary :: Buff
+
+  function DoubleBuffer{T}(::Type{Arr}, kind, dims...) where {T, Arr}
+    if kind == Host
+      secondary = zeros(FT, dims...)
+    elseif kind == Pinned
+      secondary = zeros(FT, dims...)
+      register(secondary)
+      # XXX: Should this finalizer be attached to the DoubleBuffer struct?
+      finalizer(unregister, secondary)
+    else
+      error("Bufferkind $bufferkind is not implemented yet")
+    end
+
+    buffer = new{T, typeof(primary), typeof(secondary)}(primary, secondary)
+
+    return buffer
+  end
+end
+
+register(x) = nothing
+unregister(x) = nothing
+
+@init @require CuArrays = "3a865a2d-5b23-5a0f-bc46-62713ec82fae" begin
```

Comment:
Is there a reason why we're not using `haspkg` here?

Reply:
`haspkg` is going away soon, since it is broken in 1.3